src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,352 @@
+        private MultipartPipeReaderStream _currentStream;
+        private readonly bool _canSeek;
+
+        private static ReadOnlySpan<byte> ColonDelimiter => new byte[] { (byte)':' };
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
+
+
+        public MultipartPipeReader(string boundary, PipeReader pipeReader, bool canSeek)
+        {
+            _pipeReader = pipeReader ?? throw new ArgumentNullException(nameof(pipeReader));
+            _boundary = new MultipartBoundary(boundary ?? throw new ArgumentNullException(nameof(boundary)), false);
+
+            // This stream will drain any preamble data and remove the first boundary marker. 
+            // TODO: HeadersLengthLimit can't be modified until after the constructor. 
+            _currentStream = new MultipartPipeReaderStream(_pipeReader, _boundary, canSeek);
+            _canSeek = canSeek;
+        }
+
+        /// <summary>
+        /// The limit for the number of headers to read.
+        /// </summary>
+        public int HeadersCountLimit { get; set; } = DefaultHeadersCountLimit;
+
+        /// <summary>
+        /// The combined size limit for headers per multipart section.
+        /// </summary>
+        public int HeadersLengthLimit { get; set; } = DefaultHeadersLengthLimit;
+
+        /// <summary>
+        /// The optional limit for the total response body length.
+        /// </summary>
+        public long? BodyLengthLimit { get; set; }
+
+
+        public async Task<MultipartSection> ReadNextSectionAsync(CancellationToken cancellationToken = default)
+        {
+            await _currentStream.DrainAsync(cancellationToken);
+            _bytesConsumed += _currentStream.RawLength;
+
+
+            var headersAccumulator = new KeyValueAccumulator();
+            ReadOnlySequence<byte> buffer = default;
+            ReadResult readResult = default;
+            _boundary.ExpectLeadingCrlf = true;
+            long headersLength = 0;
+            while (true)
+            {
+                if (buffer.IsEmpty)
+                {
+                    readResult = await _pipeReader.ReadAsync(cancellationToken);
+                    buffer = readResult.Buffer;
+                }
+
+                if (!buffer.IsEmpty)
+                {
+                    var finishedParsing = TryParseHeaders(ref buffer, ref headersAccumulator, ref headersLength, readResult.IsCompleted);
+                    if (headersLength > DefaultHeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
+                    }
+                    if (finishedParsing)
+                    {
+                        _bytesConsumed += headersLength;
+                        _pipeReader.AdvanceTo(buffer.Start);
+                        _currentStream = new MultipartPipeReaderStream(_pipeReader, _boundary, _canSeek);
```

Comment:
I can see why you want _canSeek for determining baseStreamOffset, but you shouldn't need to pass it into MultipartPipeReaderStream when that's not actually seekable. If something wants the individual sections to be seekable then it can add its own buffer wrapper.

Reply:
The reason I am passing _canSeek is because EnableRewind uses section.Body.CanSeek to determine if the request Body is seekable(if it's not than it doesn't buffer the file). I know that it's misleading and should be refactored - but I wanted to first create a commit that works. 

Once I address all your comments I will refactor MultipartPipeReaderStream and stop it from being a stream - When I do that I will probably have to refactor EnableRewind and the logic that buffers a FormFile or parses FormData